Source/SwiftLintFramework/Rules/Lint/DuplicateConditionsRule.swift

```diff
@@ -0,0 +1,214 @@
+        override func visitPost(_ node: IfStmtSyntax) {
+            if let prevToken = node.previousToken,
+               prevToken.tokenKind == .elseKeyword {
+                // We can skip these cases - they will be picked up when we visit the top level `if`
+                return
+            }
+
+            var maybeCurr: IfStmtSyntax? = node
+            var statementChain: [IfStmtSyntax] = []
+            while let curr = maybeCurr {
+                statementChain.append(curr)
+                maybeCurr = curr.elseBody?.as(IfStmtSyntax.self)
+            }
+
+            let positionsByConditions = statementChain
+                .reduce(into: [Set<[UInt8]>: [AbsolutePosition]]()) { acc, elt in
+                    let conditions = elt.conditions.compactMap(extract)
+                    let location = elt.conditions.positionAfterSkippingLeadingTrivia
+                    acc[Set(conditions), default: []].append(location)
+                }
+
+            addViolations(positionsByConditions)
+        }
+
+        override func visitPost(_ node: SwitchCaseListSyntax) {
+            let switchCases = node.compactMap { $0.as(SwitchCaseSyntax.self) }
+
+            let positionsByCondition = switchCases
+                .reduce(into: [[UInt8]: [AbsolutePosition]]()) { acc, elt in
+                    // Defaults don't have a condition to worry about
+                    guard case let .case(caseLabel) = elt.label else { return }
+                    for caseItem in caseLabel.caseItems {
+                        let pattern = caseItem.pattern.withoutTrivia().syntaxTextBytes
+                        let whereClause = caseItem.whereClause?.withoutTrivia().syntaxTextBytes ?? []
+                        let location = caseItem.positionAfterSkippingLeadingTrivia
+                        acc[pattern + whereClause, default: []].append(location)
+                    }
+                }
+
+            addViolations(positionsByCondition)
+        }
+
+        private func extract(_ node: ConditionElementSyntax) -> [UInt8]? {
+            let text: [UInt8]
+            switch node.condition {
+            case .availability(let node):
+                text = node.withoutTrivia().syntaxTextBytes
+            case .expression(let node):
+                text = node.withoutTrivia().syntaxTextBytes
+            case .matchingPattern(let node):
+                text = node.withoutTrivia().syntaxTextBytes
+            case .optionalBinding(let node):
+                text = node.withoutTrivia().syntaxTextBytes
+            default:
+                return nil
+            }
+
+            return text
+        }
+
+        private func addViolations<T>(_ positionsByCondition: [T: [AbsolutePosition]]) {
+            let duplicatedPositions = positionsByCondition
+                .filter { $0.value.count > 1 }
+                .flatMap { $0.value }
+
+            violations.append(contentsOf: duplicatedPositions)
```

Comment:
Perhaps a matter of taste, but the violations should only occur at real duplicates. Thus, the first condition shouldn't have a violation marker.

Reply:
I agree in principle, and originally wrote it this way. But in practice, these errors happen when there are lots of cases, and spotting the one that's wrong might be tricky without indicating both of the pair (see some of the examples picked up by SwiftLintBot). This would also be in line with `DuplicateEnumCasesRule`.

I don't have strong feelings on this at all though, and I'm very happy to change this if you reckon it's cleaner :)

Reply:
You brought a convincing argument. I'm fine with the current way of reporting on every duplicate. 👍